// react-in-viewport/dist/stories/chapters/enterCallback.stories.js
"use strict";

exports.__esModule = true;
exports["default"] = exports.FunctionalComponent = exports.ClassBaseComponent = void 0;
var _addonActions = require("@storybook/addon-actions");
var _omit = _interopRequireDefault(require("lodash/omit"));
var _index = _interopRequireWildcard(require("../../index"));
var _themeComponent = require("../common/themeComponent");
var _jsxRuntime = require("react/jsx-runtime");
var _excluded = ["forwardedRef", "inViewport"]; // eslint-disable-line import/no-named-as-default
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (e.includes(n)) continue; t[n] = r[n]; } return t; }
var ViewportBlock = (0, _index["default"])(_themeComponent.Block, {}, {
  disconnectOnLeave: false
});
var CustomAnchor = function CustomAnchor(_ref) {
  var forwardedRef = _ref.forwardedRef,
    inViewport = _ref.inViewport,
    restProps = _objectWithoutPropertiesLoose(_ref, _excluded);
  var text = inViewport ? 'Link (in viewport)' : 'Link (not in viewport)';
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("a", _extends({
    href: "https://github.com/roderickhsiao/react-in-viewport#readme"
  }, (0, _omit["default"])(restProps, _index.customProps), {
    ref: forwardedRef,
    style: {
      padding: '20px 0'
    },
    children: text
  }));
};
var ViewportAnchor = (0, _index["default"])(CustomAnchor, {}, {
  disconnectOnLeave: false
});
var _default = exports["default"] = {
  title: 'Enter Callback',
  component: ViewportBlock,
  decorators: [function (Story) {
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_themeComponent.PageTitle, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)(_themeComponent.Spacer, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        style: {
          padding: '20px',
          maxWidth: '400px'
        },
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Story, {})
      })]
    });
  }]
};
var ClassBaseComponent = exports.ClassBaseComponent = function ClassBaseComponent() {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(ViewportBlock, {
    onEnterViewport: function onEnterViewport() {
      return (0, _addonActions.action)('callback')('onEnterViewport');
    },
    onLeaveViewport: function onLeaveViewport() {
      return (0, _addonActions.action)('callback')('onLeaveViewport');
    }
  });
};
var FunctionalComponent = exports.FunctionalComponent = function FunctionalComponent() {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(ViewportAnchor, {
    onEnterViewport: function onEnterViewport() {
      return (0, _addonActions.action)('callback')('onEnterViewport');
    },
    onLeaveViewport: function onLeaveViewport() {
      return (0, _addonActions.action)('callback')('onLeaveViewport');
    }
  });
};
