// react-in-viewport/dist/stories/chapters/transition.stories.js
"use strict";

exports.__esModule = true;
exports["default"] = exports.ClassComponentTransition = void 0;
var _Section = _interopRequireDefault(require("../common/Section"));
var _themeComponent = require("../common/themeComponent");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
var _default = exports["default"] = {
  title: 'Transition',
  decorators: [function (Story) {
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_themeComponent.PageTitle, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)(_themeComponent.Spacer, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        style: {
          padding: '20px',
          maxWidth: '400px'
        },
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Story, {})
      })]
    });
  }]
};
var ClassComponentTransition = exports.ClassComponentTransition = function ClassComponentTransition() {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_Section["default"], {});
};
