// react-in-viewport/dist/next/lib/handleViewport.js
var _excluded = ["onEnterViewport", "onLeaveViewport"];
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (e.includes(n)) continue; t[n] = r[n]; } return t; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
import { useRef, forwardRef } from 'react';
import hoistNonReactStatic from 'hoist-non-react-statics';
import useInViewport from './useInViewport';
import { noop, defaultOptions, defaultConfig } from './constants';
import { jsx as _jsx } from "react/jsx-runtime";
function handleViewport(TargetComponent, options, config) {
  if (options === void 0) {
    options = defaultOptions;
  }
  if (config === void 0) {
    config = defaultConfig;
  }
  var ForwardedRefComponent = /*#__PURE__*/forwardRef(function (props, ref) {
    var refProps = {
      forwardedRef: ref
      // pass both ref/forwardedRef for class component for backward compatibility
    };
    return /*#__PURE__*/_jsx(TargetComponent, _extends({}, props, refProps));
  });
  var InViewport = function InViewport(_ref) {
    var _ref$onEnterViewport = _ref.onEnterViewport,
      onEnterViewport = _ref$onEnterViewport === void 0 ? noop : _ref$onEnterViewport,
      _ref$onLeaveViewport = _ref.onLeaveViewport,
      onLeaveViewport = _ref$onLeaveViewport === void 0 ? noop : _ref$onLeaveViewport,
      restProps = _objectWithoutPropertiesLoose(_ref, _excluded);
    var node = useRef(null);
    var _useInViewport = useInViewport(node, options, config, {
        onEnterViewport: onEnterViewport,
        onLeaveViewport: onLeaveViewport
      }),
      inViewport = _useInViewport.inViewport,
      enterCount = _useInViewport.enterCount,
      leaveCount = _useInViewport.leaveCount;
    var props = _extends({}, restProps, {
      inViewport: inViewport,
      enterCount: enterCount,
      leaveCount: leaveCount
    });
    return /*#__PURE__*/_jsx(ForwardedRefComponent, _extends({}, props, {
      ref: node
    }));
  };
  var name = TargetComponent.displayName || TargetComponent.name || 'Component';
  InViewport.displayName = "handleViewport(" + name + ")";
  return hoistNonReactStatic(InViewport, ForwardedRefComponent);
}
export default handleViewport;
