// react-in-viewport/dist/stories/common/Iframe.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;
var _react = require("react");
var _reactAspectRatio = require("react-aspect-ratio");
var _index = require("../../index");
var _jsxRuntime = require("react/jsx-runtime");
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _inheritsLoose(t, o) { t.prototype = Object.create(o.prototype), t.prototype.constructor = t, _setPrototypeOf(t, o); }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
var Iframe = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Iframe, _PureComponent);
  function Iframe(props) {
    var _this;
    _this = _PureComponent.call(this, props) || this;
    _this.state = {
      loaded: false
    };
    return _this;
  }
  Iframe.getDerivedStateFromProps = function getDerivedStateFromProps(props, state) {
    if (props.inViewport && !state.loaded) {
      return {
        loaded: true
      };
    }
    return null;
  };
  var _proto = Iframe.prototype;
  _proto.render = function render() {
    var _this$props = this.props,
      src = _this$props.src,
      ratio = _this$props.ratio;
    var Component = this.state.loaded ? 'iframe' : 'div';
    var props = this.state.loaded ? {
      src: src,
      frameBorder: 0
    } : {};
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactAspectRatio.AspectRatio, {
      ratio: ratio,
      style: {
        marginBottom: '200px',
        backgroundColor: 'rgba(0,0,0,.12)'
      },
      ref: this.props.forwardedRef,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Component, _extends({}, props))
    });
  };
  return Iframe;
}(_react.PureComponent);
var LazyIframe = (0, _index.handleViewport)(Iframe, {}, {
  disconnectOnLeave: true
});
var _default = exports["default"] = LazyIframe;
